lightwood/encoder/categorical/binary.py

```diff
@@ -1,76 +1,162 @@
-                if self.target_weights is not None:
-                    self.index_weights[self.map[word]] = 1 / self.target_weights[word]
-                else:
-                    self.index_weights[self.map[word]] = 1
 
-            self.index_weights = torch.Tensor(self.index_weights)
+            self.inv_target_weights = torch.Tensor([1, 1])  # Equally wt. both classes
+
+            # If target weights provided, weight by inverse
+            if self.target_weights is not None:
+
+                # Check target weights properly specified
+                if sum([np.isclose(i, 0) for i in self.target_weights.values()]) > 0:
+                    raise ValueError('Target weights cannot be 0')
+
+                for cat in self.map.keys():
+                    self.inv_target_weights[self.map[cat]] = (
+                        1 / self.target_weights[cat]
+                    )
 
         self.is_prepared = True
 
-    def encode(self, column_data):
+    def encode(self, column_data: Iterable[str]) -> torch.Tensor:
+        """
+        Encodes categories as OHE binary. Unknown/unrecognized classes return [0,0].
+
+        :param column_data: Pre-processed data to encode
+        :returns Encoded data of form :math:`N_{rows} x 2`
+        """  # noqa
         if not self.is_prepared:
-            raise Exception('You need to call "prepare" before calling "encode" or "decode".')
-        ret = []
+            raise Exception(
+                'You need to call "prepare" before calling "encode" or "decode".'
+            )
 
-        for word in column_data:
+        ret = torch.zeros(size=(len(column_data), 2))
+
+        for idx, word in enumerate(column_data):
             index = self.map.get(word, None)
-            ret.append([0, 0])
+
             if index is not None:
-                ret[-1][index] = 1
+                ret[idx, index] = 1
 
         return torch.Tensor(ret)
 
-    def decode(self, encoded_data, return_raw=False):
+    def decode(self, encoded_data: torch.Tensor):
+        """
+        Given encoded data, return in form of original category labels.
+        The input to `decode` makes no presumption on whether the data is already in OHE form OR not, as it some models may output a set of probabilities of weights assigned to each class. The decoded value will always be the argmax of such a vector.
+
+        In the case that the vector is all 0s, the output is decoded as "UNKNOWN"
+
+        :param encoded_data: the output of a mixer model
+
+        :returns: Decoded values for each data point
+        """  # noqa
+        encoded_data_list = encoded_data.tolist()
+        ret = []
+
+        for vector in encoded_data_list:
+            if not np.any(vector):  # Vector of all 0s -> unknown category
+                ret.append(_UNCOMMON_WORD)
+            else:
+                ret.append(self.rev_map[np.argmax(vector)])
+
+        return ret
+
+    def decode_probabilities(self, encoded_data: torch.Tensor):
```

Comment:
Adding this method without changes to `Neural` (and regression?) will result in probability predictions not working and tests failing.

Also this should return `Tuple[List[str], List[List[float]], Dict[int,str]` ... I think, or whatever the types are, not 100% sure.

Reply:
yep; I made changes in neural but this is WIP